iothub_client/tests/common_dt_e2e/iothubclient_common_dt_e2e.c

```diff
@@ -586,390 +746,238 @@ void client_create_with_properies_and_send_d2c(IOTHUB_PROVISIONED_DEVICE* device
-        )
-    {
-        if (Lock(device->lock) != LOCK_OK)
-        {
-            ASSERT_FAIL("Lock failed");
-        }
-        else
-        {
-            if ((device->receivedCallBack) && (device->cb_payload != NULL))
-            {
-                LogInfo("device->cb_payload: %s", device->cb_payload);
-
-                root_value = json_parse_string(device->cb_payload);
-                if (root_value != NULL)
-                {
-                    JSON_Object *root_object = json_value_get_object(root_value);
-                    JSON_Array *array;
-
-                    if (root_object != NULL)
-                    {
-                        switch (device->update_state)
-                        {
-                        case DEVICE_TWIN_UPDATE_COMPLETE:
-                            string_property = json_object_dotget_string(root_object, "desired.string_property");
-                            integer_property = (int)json_object_dotget_number(root_object, "desired.integer_property");
-                            array = json_object_dotget_array(root_object, "desired.array");
-                            ASSERT_IS_NOT_NULL(array, "Array not specified");
-                            integer_property_from_array = (int)json_array_get_number(array, 0);
-                            string_property_from_array = json_array_get_string(array, 1);
-
-                            break;
-                        case DEVICE_TWIN_UPDATE_PARTIAL:
-                            string_property = json_object_get_string(root_object, "string_property");
-                            integer_property = (int)json_object_get_number(root_object, "integer_property");
-                            array = json_object_get_array(root_object, "array");
-                            ASSERT_IS_NOT_NULL(array, "Array not specified");
-                            integer_property_from_array = (int)json_array_get_number(array, 0);
-                            string_property_from_array = json_array_get_string(array, 1);
-                            break;
-                        default: // invalid update state
-                            ASSERT_FAIL("Invalid update_state reported");
-                            break;
-                        }
-                        if ((string_property != NULL) && (integer_property != 0))
-                        {
-                            Unlock(device->lock);
-                            break;
-                        }
-                    }
-                }
-                json_value_free(root_value);
-                root_value = NULL;
-            }
-            Unlock(device->lock);
-        }
-        ThreadAPI_Sleep(1000);
-    }
+    _create_client_handle(device_to_use, protocol);
 
-    ASSERT_IS_TRUE(difftime(nowTime, beginOperation) < MAX_CLOUD_TRAVEL_TIME, "Timeout waiting for twin message");
+    // Twin registrations to the cloud happen asyncronously because we're using the convenience
```

Comment:
I think this isn't true anymore, because you added the block 
```
    _set_twin_callback(_twin_callback, received_twin_data);
    _receive_twin_loop(received_twin_data, DEVICE_TWIN_UPDATE_COMPLETE);
    _received_twin_data_reset(received_twin_data);
```

Because now we're not doing anything with the service SDK until we've actually setup the SUBSCRIBE and it's happily got the device SDK the twin.  ??  

If this is case we should remove commend and I think (!) we can remove the `ThreadAPI_Sleep(5000);` after this since we've made this deterministic.  Similar feedback in any other functions that Sleep() after block above (though the KillTCP message we absolutely have to Sleep() after sending)

Reply:
There is still a race condition for subscribe to twin patch topic.  We have to wait for service to finish this otherwise we can end up in an endless loop.  MQTT seems to behave ok, but AMQP takes longer to finish the subscribe.... this is where the amqp test failures were happening.

Reply:
i have rewritten the comment to reflect the updated test scneraio